### PR TITLE
Support URIs userInfo data in REST Client Reactive

### LINKED ISSUE #30289: quarkus-rest-client-reactive-jackson removing credentials from URIs and URLs
### Describe the bug

The `quarkus-rest-client-reactive-jackson` is not properly handling URIs that contain a username and password. When making a request to such a URI, the username and password are being removed from the URI, resulting in the request failing due to authentication issues.

**PD: It's also happening when using URLs:**

`RestClientBuilder.newBuilder().baseUrl(urlToInvoke)...build();`

### Expected behavior

The POST request should be made to the URI "https://USER:PASSWORD@public-develop-company-dev.cloud.com:23090/topics/topic-name" and include the username and password in the URI.

### Actual behavior

The request is made to the URI "https://public-develop-company-dev.cloud.com:23090/topics/topic-name" hence the username and password are removed from the URI

### How to Reproduce?

1. Create a new Quarkus application and add the quarkus-rest-client-reactive-jackson dependency to the project.
2. In the code, use the RestClientBuilder (`RestClientBuilder.newBuilder().baseUri(uriToInvoke)...build();`) to create a new client and set the base URI to a URI that includes a username and password, such as "https://USER:PASSWORD@public-develop-company-dev.cloud.com:23090/topics/topic-name".
3. Make a request to the URI using the created client.
4. Actual URI invoked:  "https://public-develop-company-dev.cloud.com:23090/topics/topic-name"

### Quarkus version or git rev

2.12.0.CR1 and 2.15.2.Final

### Additional information

As seen in the provided log below, got using the following application properties:

```
quarkus.rest-client.logging.scope=request-response
quarkus.rest-client.logging.body-limit=1024
quarkus.log.category."org.jboss.resteasy.reactive.client.logging".level=DEBUG
```

the quarkus-rest-client-reactive-jackson library is sending the request with an incorrect URI:

**Request:**

`POST https://public-develop-company-dev.cloud.com:23090/topics/topic-name Headers[Accept=application/json Content-Type=application/vnd.kafka.json.v2+json User-Agent=Resteasy Reactive Client content-length=407], Body:
{"records":[{"value":{
    "body": "theBody"
}}]}`

As you can see, it is not including the credentials "USER:PASSWORD" in the URI. Instead, the credentials are being removed, resulting in the request failing due to authentication issues.

Also, it can be noticed that the request contains headers such as `Accept=application/json, Content-Type=application/vnd.kafka.json.v2+json, User-Agent=Resteasy Reactive Client and content-length=407` which can also provide some information about the request.

It is important to note that this bug can cause issues when trying to connect to a service that requires authentication and can prevent the application from communicating with this service.

### PR DESCRIPTION
**Pull request by geoand**
Fixes: #30289

### CODE REVIEW

**Review by Sgitario — COMMENTED** [review]
I'm missing more context here :/ I don't follow why setting the userInfo leads to adding the Authorization header. 
Reading the issue and also checking how the Rest Client classic works, the expectation is that the userInfo should be appended to the URI:

```java
MyClient client = RestClientBuilder.newBuilder()
     .baseUri(UriBuilder.fromUri(baseUri).userInfo("foo:bar").build())
     .build(MyClient.class);
```

In Resteasy Rest Client, the resulting request is: `http://foo:bar@localhost:8081/endpoint-userinfo`. And checking the resource, I don't see any Authorization headers:
```java
@GET
    @Path("/endpoint-userinfo")
    public String callUsingUserInfo(@Context UriInfo uriInfo, @Context HttpHeaders headers) {
        // uriInfo.getRequestUri().getUserInfo() is null ??
        // headers does not contain Authorization headers
    }
```

**Comment by geoand** [comment]
@Sgitario see [this](https://learn.microsoft.com/en-us/troubleshoot/developer/browsers/security-privacy/name-and-password-not-supported-in-website-address#background-information).

Essentially when user:pass is added to an HTTP URL, most clients support converting these into basic authorization info.

**Comment by Sgitario** [comment]
> @Sgitario see [this](https://learn.microsoft.com/en-us/troubleshoot/developer/browsers/security-privacy/name-and-password-not-supported-in-website-address#background-information).
> 
> Essentially when user:pass is added to an HTTP URL, most clients support converting these into basic authorization info.
> 
> On our side, we should not have to force the user to add `.userInfo(..)` - moreover that call never comes into play when the URL is provided as a configuration property and the interface is automatically registered.
> 
> Does that answer your questions?

Mmm, I see this is a behavior for most clients, but should it not be done on the server side? I mean what I think REST Client is doing is appending the user information to the URI and sending it to the server which will extract this token to generate the authentication. It's now up to the server to accept this very insecure convention. But I think REST Client and REST Client Reactive extensions should behave the same.

**Comment by geoand** [comment]
I didn't see any information anywhere mentioning that the server should do the same thing, all information I read mentioned that the user info in the URL is turned into a basic authorization header - of course this is not secure, but that's not our problem (and can be partly mitigated when HTTPS is used) :)

**Comment by Sgitario** [comment]
> I didn't see any information anywhere mentioning that the server should do the same thing, all information I read mentioned that the user info in the URL is turned into a basic authorization header - of course this is not secure, but that's not our problem (and can be partly mitigated when HTTPS is used) :)

If you are confident about this thing, ok for me!

**Review by Sgitario — APPROVED** [review]

**Comment by geoand** [comment]
> If you are confident about this thing, ok for me!

One can never be 100% sure, but I am confident it's the right way to go :)

File: independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/AsyncInvokerImpl.java [hunk 1]
```diff
@@ -45,12 +47,30 @@ public AsyncInvokerImpl(ClientImpl restClient, HttpClient httpClient, URI uri, R
...
         this.requestSpec = new RequestSpec(requestSpec);
+        addUserInfoIfNecessary(this.uri, this.requestSpec);
         this.configuration = configuration;
         this.properties = new HashMap<>(properties);
         this.handlerChain = handlerChain;
         this.requestContext = requestContext;
     }
 
+    private void addUserInfoIfNecessary(URI uri, RequestSpec requestSpec) {
+        String userInfo = uri.getUserInfo();
```
Comment (Cs4r) [inline independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/AsyncInvokerImpl.java:58]: I have tested manually using POSTman and this is equivalent to put the user info (`USER:PASSWORD`) in the URL :white_check_mark:

File: independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/AsyncInvokerImpl.java [hunk 2]
```diff
@@ -45,12 +47,30 @@ public AsyncInvokerImpl(ClientImpl restClient, HttpClient httpClient, URI uri, R
...
+        if (userInfo == null) {
+            return;
+        }
+        String[] parts = userInfo.split(":");
+        if (parts.length != 2) {
+            return;
+        }
+        ClientRequestHeaders specHeaders = requestSpec.headers;
```
Comment (Cs4r) [inline independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/AsyncInvokerImpl.java:66]: I have tested manually using POSTman and this is equivalent to put the user info (`USER:PASSWORD`) in the URL :white_check_mark:

**Comment by Cs4r** [comment]
@geoand In which build will the changes be available?

**Comment by geoand** [comment]
It should be part of `2.16.0.Final`

**Pull request merged**